### PR TITLE
ES6 6.0 / Node v4.2.4+ Peer Review for v6.0.0

### PR DESCRIPTION
**Pull request by edwardhotchkiss**
* mostly ready, needs to return a promise always
  * checking for ES5 legacy code
  * MORE tests
  * 2 space indents on everything
  * Documentation - up to date?

If you guys have a few mins! @Jokero @niftylettuce @tsm91 @MatthewRayfield @connected-dalmer @villesau @shime - also please make sure that you are in contributors in `package.json`, thanks!

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -1,8 +1,7 @@
 language: node_js
 
 node_js:
-  - "4"
-  - "stable"
+  - "4.2.4"
```
Comment (Jokero) [inline .travis.yml]: Do you want to run tests only on nodejs 4.2.4?

File: index.js [hunk 2]
```diff
@@ -1,104 +1,96 @@
...
+    if (page !== undefined) {
+      result.page = page;
+      result.pages = Math.ceil(data.count / limit) || 1;
+    }
+    if (typeof callback === 'function') {
+      return callback(null, result);
+    }
+    let promise = new Promise();
+    promise.resolve(result);
+    return promise;
```
Comment (Jokero) [inline index.js]:
```js
return Promise.resolve(result);
```

File: tests/index.js [hunk 3]
```diff
@@ -1,179 +1,168 @@
...
-
-    it('returns promise', function() {
-        var promise = Book.paginate();
-        expect(promise.then).to.be.an.instanceof(Function);
+  });
+
+  describe('paginates', function() {
+    it('with criteria', function() {
+      return Book.paginate({ title: 'Book #10' }).then((result) => {
+        console.log('with criteria logging ==================');
```
Comment (Jokero) [inline tests/index.js]: Unnecessary?

File: index.js [hunk 4]
```diff
@@ -1,104 +1,96 @@
...
+    };
+    if (offset !== undefined) {
+      result.offset = offset;
+    }
+    if (page !== undefined) {
+      result.page = page;
+      result.pages = Math.ceil(data.count / limit) || 1;
+    }
+    if (typeof callback === 'function') {
+      return callback(null, result);
```
Comment (Jokero) [inline index.js]: Call `callback(err)` on catch

**Comment by edwardhotchkiss** [comment]
@Jokero - made the changes, need to figure out why tests are failing with some debugging later today. Thanks for the review!

File: index.js (hunk 4 above)
Comment (Jokero) [inline index.js]: I think we should not return result of callback as promise value:
```js
if (typeof callback === 'function') {
  callback(null, result);
}
return Promise.resolve(result);
```

File: index.js [hunk 5]
```diff
@@ -1,104 +1,96 @@
...
+    }
+    return Promise.resolve(result);
+  }).catch((error) => {
+    return callback(error);
```
Comment (Jokero) [inline index.js]: So better?
```js
if (typeof callback === 'function') {
  callback(error);
  return Promise.reject(error);
}
```

**Comment by edwardhotchkiss** [comment]
@Jokero Check it out, all tests passing. Looks pretty sexy. We can make some code smaller but it seems to be working great now. Thoughts on release as `v6.0.0`?

File: index.js [hunk 6]
```diff
@@ -18,13 +20,13 @@
 
 function paginate(query, options, callback) {
   query = query || {};
-  options = Object.assign({}, paginate.options, options);
-  let select = options.select;
+  options = options || {};
```
Comment (Jokero) [inline index.js]: Why did you remove `paginate.options`?

File: index.js [hunk 7]
```diff
@@ -18,13 +20,13 @@
...
+  let select = options.select || null;
   let sort = options.sort;
   let populate = options.populate;
   let lean = options.lean || false;
-  let leanWithId = options.leanWithId ? options.leanWithId : true;
-  let limit = options.limit ? options.limit : 10;
+  let leanWithId = options.leanWithId || false;
```
Comment (Jokero) [inline index.js]: Why not `true `?

File: index.js [hunk 8]
```diff
@@ -37,58 +39,66 @@ function paginate(query, options, callback) {
...
-    };
-    if (lean && leanWithId) {
-      promises.docs = promises.docs.then((docs) => {
-        docs.forEach((doc) => {
-          doc.id = String(doc._id);
-        });
-        return docs;
-      });
-    }
+  let docsQuery = this.find(query)
```
Comment (Jokero) [inline index.js]: What if limit is zero?

File: index.js (hunk 6 above)
Comment (edwardhotchkiss) [inline index.js]: @Jokero There was nothing ever sent into paginate.options and that which was wasn't being tested. It could be added if we had it in the docs and tests for it ... feel free to push :)

File: index.js (hunk 7 above)
Comment (edwardhotchkiss) [inline index.js]: Most people don't need an "id" specifically, they just need { "lean" : true }, and to access _id, whether it is id or _id. Thoughts?

File: index.js (hunk 8 above)
Comment (edwardhotchkiss) [inline index.js]: Limit is specified below, want to push a test to show me what you mean?

File: index.js (hunk 6 above)
Comment (Jokero) [inline index.js]: paginate.options, limit=0, leanWithId=true were added in 5.0.0, tested and documented

https://github.com/edwardhotchkiss/mongoose-paginate/tree/v5.0.0#set-custom-default-options-for-all-queries
https://github.com/edwardhotchkiss/mongoose-paginate/tree/v5.0.0#zero-limit

File: index.js (hunk 8 above)
Comment (Jokero) [inline index.js]: https://github.com/edwardhotchkiss/mongoose-paginate/blob/v5.0.0/tests/index.js#L128

File: index.js (hunk 7 above)
Comment (Jokero) [inline index.js]: By default mongoose add `id` field to every document. So if you use it and need `lean` you can get `id` without any special actions

**Comment by edwardhotchkiss** [comment]
@Jokero I added back paginate.options, the id field is back. One issue left is the `limit : 0`. The test was passing because of how the query was built. I think that no one will ever use limit as zero. Thoughts? Otherwise I think that we're good to go. :8ball:

**Comment by edwardhotchkiss** [comment]
Also all tests passing except with limit zero.

**Comment by Jokero** [comment]
I use `limit: 0` because it's useful in rest api to get only metadata (page, total, but not documents). Maybe anyone else uses this functionality

**Comment by edwardhotchkiss** [comment]
Yeah limit:0 in MongoDB defaults to all docs as the default behavior. I'll make sure that there is a note of it in the docs.

**Comment by Jokero** [comment]
Pagination is fetching and counting. If we set `limit: 0` we disable fetching. It's very cool

**Comment by edwardhotchkiss** [comment]
@Jokero limit:0 back to original ... :8ball:

**Comment by antony** [comment]
Is this due to be merged sometime soon? 

Reason being is that I want to start adding a feature but I don't want to do it on the old codebase  and end up in merge hell :)

**Comment by simison** [comment]
@antony see #93

**Comment by antony** [comment]
Ah that's terrible news. Sorry to hear it.

@simison I can have a look at getting this PR finalised, I'm quite time constrained so I don't know how soon I'll be able to do it, but I'll update this ticket as I progress.

File: index.js [hunk 9]
```diff
@@ -37,58 +40,69 @@ function paginate(query, options, callback) {
...
+
+  return promise;
 }
 
 /**
  * @param {Schema} schema
+ * use native ES6 promises verus mpromise
  */
 
+mongoose.Promise = global.Promise;
```
Comment (fiznool) [inline index.js]: I would be wary about this line. Some people like to set `mongoose.Promise` to their own promise implementation (e.g. bluebird). Overwriting it here would cause them lots of headaches.
